### PR TITLE
Add TypeScript template

### LINKED ISSUE #2: Typescript
a typescript template version would be awesome!

### PR DESCRIPTION
**Pull request by LitoMore**
Resolve #2 

### Difference from #4 

- Use `--typescript` option for creating with TypeScript React
- Organize template files, common files have been moved to `templates/_common` directory
- Use `@sindresorhus/tsconfig`
- `source` directory name for source code
- Has exports - Component compatible
- My PR is alive :)

### CODE REVIEW

**Comment by LitoMore** [comment]
OMG, I just realized my PR is repeated. 🤦‍♂️

**Comment by LitoMore** [comment]
BTW, I can add the component creating support on another PR.

```shell
$ create-ink-app --component
# or
$ craete-ink-component
```

File: cli.js [hunk 1]
```diff
@@ -4,6 +4,9 @@ const meow = require('meow');
 const createInkApp = require('.');
 
 meow(`
+	Options
+		--tsx		Use TypeScript React template
```
Comment (vadimdemedes) [inline cli.js:8]:
```suggestion
		--typescript		Use TypeScript React template
```

File: index.js [hunk 2]
```diff
@@ -22,9 +22,65 @@ const copyWithTemplate = async (from, to, variables) => {
 	await writeFile(to, generatedSource);
 };
 
-const fromPath = file => path.join(__dirname, 'template', file);
+const useTsx = process.argv.includes('--tsx');
```
Comment (vadimdemedes) [inline index.js:25]:
```suggestion
const useTypeScript = process.argv.includes('--tsx');
```

File: index.js [hunk 3]
```diff
@@ -22,9 +22,65 @@ const copyWithTemplate = async (from, to, variables) => {
...
+let templatePath = 'templates/js';
+
+if (useTsx) {
+	templatePath = 'templates/tsx';
```
Comment (vadimdemedes) [inline index.js:29]:
```suggestion
	templatePath = 'templates/ts';
```

Could you rename the folder too?

File: index.js [hunk 4]
```diff
@@ -22,9 +22,65 @@ const copyWithTemplate = async (from, to, variables) => {
...
+}
+
+const fromPath = file => path.join(__dirname, templatePath, file);
 const toPath = file => path.join(process.cwd(), file);
 
+const copyTasks = variables => {
+	const commonCopies = [
```
Comment (vadimdemedes) [inline index.js:36]:
```suggestion
	const commonTasks = [
```

**Review by vadimdemedes — CHANGES_REQUESTED** [review]

**Comment by vadimdemedes** [comment]
Have some minor comments, but great work overall!

> BTW, I can add the component creating support on another PR.

I think it'd better fit in a `create-ink-component` package instead.

**Comment by vadimdemedes** [comment]
Hey @LitoMore, I'm getting this error when running `npm test` in a TypeScript CLI:

```
> xo && ava


  source/cli.tsx:6:13
  ✖  6:13  A space is forbidden before closing bracket  react/jsx-tag-spacing

  1 error
npm ERR! Test failed.  See above for more details.
```

**Comment by LitoMore** [comment]
> Hey @LitoMore, I'm getting this error when running npm test in a TypeScript CLI

Fixed.

BTW, I can also enhance our CI to test the generated app. Should I add the generated app test in this PR or open a new PR?

**Comment by vadimdemedes** [comment]
Good idea, let's do it in a separate PR ;)

**Comment by LitoMore** [comment]
> let's do it in a separate PR ;)

OK, now this PR is done.

**Comment by vadimdemedes** [comment]
There's a regression, TS version doesn't have `--help` support and doesn't even use `meow`. See `cli.js` in JS version.

**Comment by LitoMore** [comment]
> There's a regression, TS version doesn't have --help support and doesn't even use meow. See cli.js in JS version.

Resolved.

File: index.js [hunk 5]
```diff
@@ -12,6 +13,9 @@ const readFile = promisify(fs.readFile);
 const writeFile = promisify(fs.writeFile);
 
 const copyWithTemplate = async (from, to, variables) => {
+	const dirname = path.dirname(to);
+	await makeDir(dirname);
```
Comment (vadimdemedes) [inline index.js:17]: Why is this needed now?

File: templates/ts/source/ui.tsx [hunk 6]
```diff
@@ -1,7 +1,7 @@
-import React from 'react';
+import React, {FC} from 'react';
 import {Text} from 'ink';
 
-const App = ({name = 'Stranger'}) => (
+const App: FC<{name?: string}> = ({name = 'Stranger'}) => (
```
Comment (vadimdemedes) [inline templates/ts/source/ui.tsx:4]: Could you separate that into an interface?

```ts
interface Props {
  name: string | undefined
}

const App: FC<Props> = ...
```

File: index.js (hunk 5 above)
Comment (LitoMore) [reply index.js:17]: `copyWithTemplate` is not isuing `cpy` for copying files.

`cpy` can make directory automatically, but the `copyWithTemplate` is not.

File: templates/ts/source/ui.tsx (hunk 6 above)
Comment (LitoMore) [reply templates/ts/source/ui.tsx:4]: Sure.

**Comment by vadimdemedes** [comment]
Great work, thanks @LitoMore!
